File: db4oj.tests/src/com/db4o/db4ounit/common/internal/TransactionLocalTestCase.java
package com.db4o.db4ounit.common.internal;

import com.db4o.internal.*;

import db4ounit.*;
import db4ounit.extensions.*;

public class TransactionLocalTestCase extends AbstractInMemoryDb4oTestCase {
	
	static class Item {
		
		public final Transaction transaction;

		public Item(Transaction transaction) {
			this.transaction = transaction;
		}
	}
	
	private final TransactionLocal<Item> _subject = new TransactionLocal<Item>() {
		@Override
		public Item initialValueFor(Transaction transaction) {
			return new Item(transaction);
		}
	};
	
	private Transaction t1;
	private Transaction t2;
	
	@Override
	protected void db4oSetupAfterStore() throws Exception {
		t1 = newTransaction();
		t2 = newTransaction();
	}
	
	public void testValueRemainsTheSame() {
		Assert.areSame(itemFor(t1), itemFor(t1));
		Assert.areSame(itemFor(t2), itemFor(t2));
	}
	
	public void testDifferentValuesForDifferentTransactions() {
		Assert.areNotSame(itemFor(t1), itemFor(t2));
	}
	
	public void testInitialValueTransaction() {
		Assert.areSame(t1, itemFor(t1).transaction);
		Assert.areSame(t2, itemFor(t2).transaction);
	}
	
	public void testValuesAreDisposedOfOnCommit() {
		final Item itemBeforeCommit = itemFor(t1);
		t1.commit();
		final Item itemAfterCommit = itemFor(t1);
		Assert.areNotSame(itemAfterCommit, itemBeforeCommit);
	}
	
	public void testValuesAreDisposedOfOnRollback() {
		final Item itemBeforeRollback = itemFor(t1);
		t1.rollback();
		final Item itemAfterRollback = itemFor(t1);
		Assert.areNotSame(itemAfterRollback, itemBeforeRollback);
	}

	private Item itemFor(final Transaction transaction) {
	    return transaction.get(_subject).value;
    }

}

File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Internal/TransactionLocalTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using Db4oUnit;
using Db4oUnit.Extensions;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Tests.Common.Internal;

namespace Db4objects.Db4o.Tests.Common.Internal
{
	public class TransactionLocalTestCase : AbstractInMemoryDb4oTestCase
	{
		internal class Item
		{
			public readonly Transaction transaction;

			public Item(Transaction transaction)
			{
				this.transaction = transaction;
			}
		}

		private sealed class _TransactionLocal_19 : TransactionLocal
		{
			public _TransactionLocal_19()
			{
			}

			public override object InitialValueFor(Transaction transaction)
			{
				return new TransactionLocalTestCase.Item(transaction);
			}
		}

		private readonly TransactionLocal _subject = new _TransactionLocal_19();

		private Transaction t1;

		private Transaction t2;

		/// <exception cref="System.Exception"></exception>
		protected override void Db4oSetupAfterStore()
		{
			t1 = NewTransaction();
			t2 = NewTransaction();
		}

		public virtual void TestValueRemainsTheSame()
		{
			Assert.AreSame(ItemFor(t1), ItemFor(t1));
			Assert.AreSame(ItemFor(t2), ItemFor(t2));
		}

		public virtual void TestDifferentValuesForDifferentTransactions()
		{
			Assert.AreNotSame(ItemFor(t1), ItemFor(t2));
		}

		public virtual void TestInitialValueTransaction()
		{
			Assert.AreSame(t1, ItemFor(t1).transaction);
			Assert.AreSame(t2, ItemFor(t2).transaction);
		}

		public virtual void TestValuesAreDisposedOfOnCommit()
		{
			TransactionLocalTestCase.Item itemBeforeCommit = ItemFor(t1);
			t1.Commit();
			TransactionLocalTestCase.Item itemAfterCommit = ItemFor(t1);
			Assert.AreNotSame(itemAfterCommit, itemBeforeCommit);
		}

		public virtual void TestValuesAreDisposedOfOnRollback()
		{
			TransactionLocalTestCase.Item itemBeforeRollback = ItemFor(t1);
			t1.Rollback();
			TransactionLocalTestCase.Item itemAfterRollback = ItemFor(t1);
			Assert.AreNotSame(itemAfterRollback, itemBeforeRollback);
		}

		private TransactionLocalTestCase.Item ItemFor(Transaction transaction)
		{
			return ((TransactionLocalTestCase.Item)transaction.Get(_subject).value);
		}
	}
}
